walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,67 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList = new ArrayList(events);
+
+    Collections.sort(eventList, new Comparator<Event>() {
+      @Override
+      public int compare(Event event1, Event event2) {
+        return event1.getWhen().start() - event2.getWhen().start();
+      }
+    });
+
+    int currentRangeStart = TimeRange.START_OF_DAY;
+    int currentRangeEnd;
+    int currentRangeDuration;
+
+    boolean sharesAttendees;
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+
+    for (Event event : eventList) {
+      sharesAttendees = false;
+
+      for (String person : event.getAttendees()) {
+        if (request.getAttendees().contains(person)) {
+          sharesAttendees = true;
+          break;
+        }
+      }
+
+      if (!sharesAttendees) continue;
+
+      currentRangeEnd = event.getWhen().start();
+      currentRangeDuration = currentRangeEnd - currentRangeStart;
+
+      if (currentRangeEnd >= currentRangeStart) {
+        if (currentRangeDuration >= request.getDuration()) {
+          availableTimes.add(
+            TimeRange.fromStartDuration(currentRangeStart, currentRangeDuration));
+        }
+        currentRangeStart = event.getWhen().end();
+      } else {
+        if (event.getWhen().end() > currentRangeStart) {
+          currentRangeStart = event.getWhen().end();
+        }
+      }
+    }
+
+    currentRangeEnd = TimeRange.END_OF_DAY;
+    currentRangeDuration = currentRangeEnd - currentRangeStart;
+    if (currentRangeDuration >= request.getDuration()) {
```

Comment:
Maybe just skip the currentRangeDuration assignment above and just check 'if (currentRangeEnd - currentRangeStart >= request.getDuration())'?

Reply:
Done